src/olympia/files/tests/test_models.py

```diff
@@ -1065,6 +1065,36 @@ def test_webextension_crx_large(self):
         assert upload.path.endswith('.xpi')
         storage.delete(upload.path)
 
+    def test_webextension_crx_version_3(self):
+        """Test to ensure we accept CRX uploads (version 3), but convert them
+        into XPI files ASAP to keep things simple.
+        """
+        upload = self.get_upload('webextension_crx3.crx')
+        assert upload.path.endswith('.xpi')
+        storage.delete(upload.path)
+
+    def test_webextension_crx_not_a_crx(self):
+        """Test to ensure we raise an explicit exception when a .crx file isn't
+        a true crx (doesn't have to be caught, showing a 500 error is fine)."""
+        data = b'Cr42\x02\x00\x00\x00&\x01\x00\x00\x00\x01\x00\x00'
+        with self.assertRaises(InvalidOrUnsupportedCrx):
+            FileUpload.from_post([data], filename='test.crx', size=1234)
+
+    def test_webextension_crx_version_unsupported(self):
+        """Test to ensure we only support crx versions 2 and 3 and raise an
+        explicit exception otherwise (doesn't have to be caught, showing a 500
+        error is fine)."""
+        data = b'Cr24\x04\x00\x00\x00&\x01\x00\x00\x00\x01\x00\x00'
+        with self.assertRaises(InvalidOrUnsupportedCrx):
+            FileUpload.from_post([data], filename='test.crx', size=1234)
+
+    def test_webextension_crx_version_cant_unpack(self):
+        """Test to ensure we raise an explicit exception when we can't unpack
+        a crx (doesn't have to be caught, showing a 500 error is fine)."""
+        data = b'Cr24\x02\x00\x00\x00&\x00\x00\x00\x01\x00\x00'
```

Comment:
Can you differentiate the exceptions theses blobs of data are raising?  They could all be just bad data raising in the same place rather than good data that's the wrong version, etc, raising for different reasons.

Reply:
Yeah I can look in the message, I'll do that.

Reply:
c1cf299ed8e2250d3584ed2e493acba14a8e57af